### DISCUSSION TITLE
XState to interface with GraphQL

### DISCUSSION THREAD

**Discussion by tjdam** [statelyai/xstate#1349] [General] [answered]
Hi,

I'm new to frontend and state management (with a library) in React so this might be a silly question but, I started writing some XState actions that deals directly with GraphQL mutations and started liking it more than having them in React Components. Even thought of moving all queries and mutations to actions and let XState handle everything.

But I ~think~ I remember reading someone saying that XState should only handle my app state and not be that tightly coupled with my backend API.

Thoughts?

**Comment by davidkpiano** [comment, marked as answer]
> But I think I remember reading someone saying that XState should only handle my app state and not be that tightly coupled with my backend API.

Hm, well _I_ definitely didn't say that 😅 

What do your GraphQL mutations look like in XState? Can you share some sample code?

**Reply by tjdam** [reply]
Hahah 😅
I'm trying to think where I got this from!

But anyways, I have for instance this simple helper function that I use in a couple of action to updates a value (a boolean) in the database, and it looks like this:
```js
const completeHelper = async (id, complete) => {
  const client = initializeApollo()

  return await client.mutate({
    mutation: gql`
    mutation UpdateStatus($input: InputType) {
      result: updateStatus(input: $input) {
          id
       }
     }
    `,
    variables: {
      input: { id, complete } 
    })
}


// actions
export const completeFalse = (_ctx, e) => {
  completeHelper(e.contentId, false)
}

export const completeTrue = (_ctx, e) => {
  completeHelper(e.contentId, true)
}
```

This was somenthing that was appearning in my app in more than one place and now it's just a matter of calling this action when a given XState event occurs, which I find really neat :)

Do you see any problem with this?

**Reply by davidkpiano** [reply]
It's fine as-is, but keep in mind, that as an action it is **fire-and-forget**, which means you don't care about the response.

If you do care about the response, you can invoke this instead.

**Reply by tjdam** [reply]
Meaning, I shouldn't use it to handle the response from these mutations? Why is that?

**Reply by Andarist** [reply]
@davidkpiano has meant that actions are not that well-suited for side-effects that return something or for ones that are async and your `completeHelper` is both.

Technically you could just grab a reference to your service and `.send()` to it from within your action, but a better alternative is to use invoked services to have a tight bidirectional communication channel between your parent and a child. It also has a better visualization support because if you just `.send()` from within your action we don't know about it at all, so, at best, sent event can be represented as a random~ external event, without contextual meaning. Whereas with invoking services it becomes much clearer on the visualized graph that a particular event is a "response" from an invoked child.

**Reply by tjdam** [reply]
Thanks for the answer @davidkpiano and @Andarist!
And sorry for the late reply, I went down a rabbit hole looking into invoke. I skipped this before cause it was presented in the context of the actor model and I thought it was too much for me as I was still trying to wrap my head around state machines.

This is really really neat. 
Thanks for your work!

Edit: in case someone ends up here with the same question, I just found David's answer to Scott Tolinski on that and it's exactly what I wanted!: https://codesandbox.io/s/xstate-apollo-client-u8is1?file=/src/index.js

**Reply by blocka** [reply]
A downside of this approach is that you can't take advantage of apollo's update cache on mutation feature.

**Reply by davidkpiano** [reply]
@blocka Can you show a code example of what you're talking about?

**Reply by blocka** [reply]
Sure,
Imagine we have a simple schema

```graphql
type Counter {
    id: ID!
    count: Int!
}

type Query {
  getCounter: Counter!
}

type Mutation {
  increment: Counter!
}
```

if we have a component which looks something like this:

```js
const GET_COUNTER = gql`
   query GetCounter {
     getCounter {
       id
       count
     }
   }
`

const INCREMENT = gql`
  mutation Increment {
    increment {
      id
      count
    }
  }
`
function Counter () {
  const { data } = useQuery(GET_COUNTER)
  const [increment] = useMutation(INCREMENT)

  return (
    <div>
      Counter is at {data?.getCounter?.count}

      <button onClick={increment}>+1</button>
    </div>
  )
}
```

when the data from this mutation comes back, apollo will update it's cache, and the component will rerender

I think what's basically missing here, is that this is working here with an observable, not just a simple promise.

So perhaps it would be a better idea to invoke an observable by calling `client.watchQuery` instead of `client.query`.

I imagine that could also be abstracted into it's own reusable machine as well.

**Reply by davidkpiano** [reply]
That makes sense, I'll try to get some Apollo + GraphQL examples with mutations up soon!

**Reply by tjdam** [reply]
> That makes sense, I'll try to get some Apollo + GraphQL examples with mutations up soon!

Out of curiosity, did you end up writing and posting this example somewhere? 😅

**Reply by davidkpiano** [reply]
@tjdam Mutations would be done using the same concepts, except Apollo Client makes it awkward because it forces (?) you to use a hook:

```js
const [login, { loading, error }] = useMutation(/* ... */);
const [state, send] = useMachine(loginMachine, {
  actions: { login: () => login(/* .. */) }
});

// ...
```

**Reply by UberMouse** [reply]
I make use of GraphQL extensively in a fully XState powered UI. I can use cache updaters fine. You do need to ensure that all your queries are Observables by using `apolloClientInstance.watchQuery` instead of `query`. I structure my machines with two top level parallel states, `ui` and `data`. The data state invokes an observable  that emits GraphQL query results and those emits gets transformed into xstate events which I use to transition through states/update the machines context. It's invoked into its own state so it sticks around as the machine transitions through ui states (though I think I just realized I could move this to a top level invoke and get the same effect...).

I've built up a number of helpers to assist in this, I've ripped out some examples from our codebase to demonstrate, they are not cleaned up and are missing some utilities/imports etc but hopefully it gives you an idea of how to do it https://gist.github.com/UberMouse/62631a290a7fbbb6759ffa0344d68039

**Reply by blocka** [reply]
Yeah...I think that's the key...invoking it as an observable, not as a promise. This way a mutation can continue to be "fire-and-forget", but the cache will still be updated (and this will be true for any other machines which subscribe to queries that should be updated by the mutation)

**Reply by veeramarni** [reply]
@UberMouse Does it work with latest apollo client?

**Reply by UberMouse** [reply]
We use @apollo/client v3.7 now so it should. I can't speak to whether it required some updates for v3 compat, pretty sure it either didn't or was minor.

**Reply by veeramarni** [reply]
@UberMouse can you please share full script for `createDataStream` as it is not working for me.
https://gist.github.com/UberMouse/62631a290a7fbbb6759ffa0344d68039?permalink_comment_id=4459957#gistcomment-4459957

**Reply by veeramarni** [reply]
I figured, Observable from rxjs made it work.
